### PR TITLE
Convert & consolidate protocol types in Kotlin

### PR DESCRIPTION
**Pull request by Strum355**
As described in the title. Reviewing on Github might be less optimal due to the order of changed files, I've found that the following is very convenient:
1. checkout the branch locally (`git fetch && git checkout nsc/protocol-kt`)
2. load the changes into the staging area (`git reset --soft HEAD~1`)
3. in the vscode "Source control" view, sort the changes by (file) name (see screenshot at the very bottom[1])

Some notes:
1. The two ChatMessage types were collapsed into one, the [issue described regarding the enum serialization](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-9c533c092b1847a0028f4f0b8d02ceab7a330c8805b5b4df79fe45a012b33d3bL6-L7) appears to be [handled by `@SerializedName`](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-c514334703915e2ca2b644739ef330a3993bd9745c2fc27c0b444b22d38f3c5aR6-R7)
2. Some field types were updated to reflect the current state of the types in the https://github.com/sourcegraph/cody repo, which saves us some null checking in some areas (e.g. [title and id on `RecipeInfo`](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-335702a83e5c6967876175457e9f02c4c87ccd4b2b6bde508476c6b3f197092fL249-L251)
3. There are some TODOs that are best done when further classes are converted to Kotlin e.g. [the `Event` type (which is in need of updating)](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-135b4c57ef38eece3f8fb3ee1ada17a0fd807a1382f1789d311b0308cb2faa37R8), [a `@JvmOverloads` on `TextDocument`](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-15b5c585ca94aa06d1bca657c19df89aa1e10f6e186dd12f9f3c5c36eb6fbc46R4-R5)
4. Further protocol types are not yet introduced here, such as `RecipeID` required [here](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-8d082a76e3f9735bc2c5e134b233a72984669e3004ada4487781becaa5a7b8ecR4) and [here](https://github.com/sourcegraph/jetbrains/compare/nsc/protocol-kt?expand=1#diff-8ed923530b9e0f875871757a03cf8c6764a83a47445f21f6eb036de0aa47c7fdR4)
5. more to be filled out as I remember

<details><summary>Screenshot [1]</summary>

image

</details> 

## Test plan

Code type-checks & compiles, chat & autocomplete still working from local testing, no gson or agent errors encountered

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 1]
```diff
@@ -246,13 +247,10 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
...
     for (recipe in recipes) {
-      if (recipe.id == null || recipe.title == null) {
-        continue
-      }
-      val recipeButton = createRecipeButton(recipe.title!!)
+      val recipeButton = createRecipeButton(recipe.title)
       recipeButton.addActionListener {
         GraphQlLogger.logCodyEvent(project, "recipe:" + recipe.id, "clicked")
-        sendMessage(project, recipe.title!!, recipe.id!!)
+        sendMessage(project, recipe.title, recipe.id)
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt:253]: I see that a few fields are no longer nullable thanks to these changes. this is great 🚀

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/ChatMessage.kt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.sourcegraph.cody.agent.protocol
+
+data class ChatMessage(
+    override val speaker: Speaker,
+    override val text: String?,
+    val displayText: String? = null,
+    val contextFiles: List<ContextFile>? = null
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/ChatMessage.kt:7]: does it make sense to distinguish `contextFiles`: `null` and `emptyList`?

File: src/main/java/com/sourcegraph/cody/agent/protocol/ChatMessage.java [hunk 3]
```diff
@@ -1,33 +0,0 @@
-package com.sourcegraph.cody.agent.protocol;
-
-import java.util.List;
-import org.jetbrains.annotations.Nullable;
-
-// TODO: consolidate with the other ChatMessage. This duplication exists because the other
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/cody/agent/protocol/ChatMessage.java:6]: nice. this is done btw! 🚀

**Review by mkondratek — APPROVED** [review]
Reviewed! LGTM!

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/ChatMessage.kt (hunk 2 above)
Comment (Strum355) [reply src/main/kotlin/com/sourcegraph/cody/agent/protocol/ChatMessage.kt:7]: Just mirroring how its defined in the cody repo. Im not sure how deserializing `null` works into a property with a default value e.g. if this was `val contextFiles: List<ContextFile> = emptyListOf()` :thinking:

**Pull request merged**